### DISCUSSION TITLE
I ran the scrolling example, but the screen seemed to have a residual image and refresh slowly

### DISCUSSION THREAD

**Discussion by ShiPC123** [board707/DMD_STM32#110] [General]
I used STM32F103RCT6 and F411 microcontroller, the display effect is the same

https://github.com/board707/DMD_STM32/assets/132763025/beed7302-e149-4535-a0e7-0d6038bccf7f

Here is my test code
void loop(void)
{

    // create foreground colors
    uint16_t col[] = {
        dmd.Color888(255,0, 0), // red

    };
    uint16_t bg = 0;  // background - black
    int col_cnt = 3;   // color count

    // text
    char s[] = "hello 1122334455";
    // transcode message to UTF for use with GFX fonts
    char k[30];
    
    utf8_rus(k, (const unsigned char*)s);
    char* m = s;
    // select standard font
    dmd.selectFont(&UkrRusArial_F);
    
    // set text foreground and background colors
    dmd.setTextColor(col[0], bg);

    // shift steps in pixels for running text (positive - shift right, negative - left)
    int8_t step[] = { -1,-1,-1,-1 };
    // running text shift interval
    uint16_t interval = 10;

    long prev_step = millis();
    uint8_t col_ptr = 0;
    uint8_t i = 0, b = 0;
    uint8_t test = 255;
    uint8_t test_cnt = 4;
    dmd.setBrightness(100);
    int x=dmd.width() - 1;
    while (1) {
        if ((millis() - prev_step) > interval) {
          prev_step = millis();  // moved to the start of the loop to prevent "scroll not smooth" issue
          x--;
          if(x<-1*dmd.stringWidth(m))
          {
            x= dmd.width() - 1;
          }

        }
          dmd.drawMarqueeX(m, x, 12);
          dmd.swapBuffers(true);
        
    }
}

**Comment by ShiPC123** [comment]
https://github.com/board707/DMD_STM32/assets/132763025/8b303c3e-6ce0-48e4-83bc-6b094e4f433d

**Comment by board707** [comment]
Hi
Could you clarify an issue? I didn't catch the problem from the video.

**Reply by ShiPC123** [reply]
Hello, maybe can't see the effect from the inside of the camera, but to the naked eye,  can see how Carton feels. The rolling is not smooth

**Reply by ShiPC123** [reply]
How can I improve the flow of the roll

**Comment by board707** [comment]
Honesty, I don't see any distortions :)
I have try it at home.

By anyway, try to increase a scrolling interval:
`uint16_t interval = 10;`
The 10 mS is probably too tight for the task.

**Comment by ShiPC123** [comment]
Thanks, I will try to re-shoot the video tomorrow, and I feel that the shadow will appear when I scroll

**Comment by ShiPC123** [comment]
Is there any relevant document that can explain the function, such as dmd.drawMarqueeX? I wonder if I didn't clear the screen or didn't call some functions, which caused the scrolling to have residual shadows

**Comment by board707** [comment]
Read the library [Wiki](https://github.com/board707/DMD_STM32/wiki)
The brief explanation of the functions: [DMD_STM32 Reference](https://github.com/board707/DMD_STM32/wiki/Reference)

**Comment by Suresh6060** [comment]
I have used this library with very good success. It is true , I had to struggle a bit to get it running though. 

There are some quirks that I could not understand fully. May be it is timing or memory management issue. 

In my case , I had to declare the variable m[] as a global variable. This introduced new problems with unknown characters or characters of the previous scrolls added to the new scroll. Technically this should not happen. But that was the case. To solve the issue I add a for loop and re-assigned m[]=NULL  , resetting the variable , before using it again . BTW i was dynamically populating m[] and not sending a static text to drawMarquee (). 

The variable interval practically determines the speed to scroll. 

I this this library works in two stage. drawMarquee had to be initially setup as it initializing with stepmarquivee etc. don't use swapbuffers() here. 

The second stage is used to subsequent repeat, reversal, vertical etc. 

This is a little bit different from how DMD is implemented in esp32 etc. 

I might be wrong, but this is what I understood and used. 

At one point I had the slowing issue and tested outputing  the actual time  each section of code takes to finish and tweeked the code. Hope you know C is an admant , stuborn and unforgiving langauge. memory management has to very precise.

**Reply by board707** [reply]
> I had to declare the variable m[] as a global variable. This introduced new problems with unknown characters or characters of the previous scrolls added to the new scroll. Technically this should not happen. But that was the case.

Did you forget to add the string terminator at the end of the line?

For this and another issues, please show the code that demonstrate the problem.
